server/src/http/broadcaster.rs

```diff
@@ -0,0 +1,207 @@
+/// copied from https://github.com/actix/examples/blob/master/server-sent-events/src/broadcast.rs
+use std::{
+    hash::{Hash, Hasher},
+    sync::Arc,
+    time::Duration,
+};
+
+use actix_web::{
+    rt::time::interval,
+    web::{Json, Query},
+};
+use actix_web_lab::{
+    sse::{self, Event, Sse},
+    util::InfallibleStream,
+};
+use dashmap::DashMap;
+use futures::future;
+use serde::Serialize;
+use tokio::sync::mpsc;
+use tokio_stream::wrappers::ReceiverStream;
+use unleash_types::client_features::{ClientFeatures, Query as FlagQuery};
+
+use crate::{
+    filters::{filter_client_features, name_prefix_filter, project_filter, FeatureFilterSet},
+    tokens::cache_key,
+    types::{EdgeJsonResult, EdgeResult, EdgeToken, FeatureFilters},
+};
+
+#[derive(Debug, Clone, PartialEq, Eq, Serialize)]
+struct QueryWrapper {
+    query: FlagQuery,
+}
+
+impl Hash for QueryWrapper {
+    fn hash<H: Hasher>(&self, state: &mut H) {
+        serde_json::to_string(&self.query).unwrap().hash(state);
+    }
+}
+
+#[derive(Clone, Debug)]
+struct ClientGroup {
+    clients: Vec<mpsc::Sender<sse::Event>>,
+    filter_set: Query<FeatureFilters>,
+    token: EdgeToken,
+}
+
+pub struct Broadcaster {
+    active_connections: DashMap<QueryWrapper, ClientGroup>,
+    features_cache: Arc<DashMap<String, ClientFeatures>>,
+}
+
+impl Broadcaster {
+    /// Constructs new broadcaster and spawns ping loop.
+    pub fn new(features: Arc<DashMap<String, ClientFeatures>>) -> Arc<Self> {
+        let this = Arc::new(Broadcaster {
+            active_connections: DashMap::new(),
+            features_cache: features,
+        });
+
+        Broadcaster::spawn_ping(Arc::clone(&this));
+
+        this
+    }
+
+    /// Pings clients every 30 seconds to see if they are alive and remove them from the broadcast
+    /// list if not.
+    fn spawn_ping(this: Arc<Self>) {
+        tokio::spawn(async move {
+            let mut interval = interval(Duration::from_secs(30));
+
+            loop {
+                interval.tick().await;
+                this.remove_stale_clients().await;
+            }
+        });
+    }
+
+    /// Removes all non-responsive clients from broadcast list.
+    async fn remove_stale_clients(&self) {
```

Comment:
I see this method's main responsibility as the "heartbeat". The fact that it removes non-sendable clients from the list feels like a detail. Like in Unleash's implementation.

```suggestion
    async fn heartbeat(&self) {
```

Reply:
Yeah, I can agree with that. Renamed for now.